File: android/app/src/main/java/com/donetick/app/widget/WidgetBridgePlugin.java
package com.donetick.app.widget;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * JS bridge for the home-screen widgets (see src/service/WidgetService.js).
 * Persists the task snapshot + API config and redraws the widgets.
 */
@CapacitorPlugin(name = "WidgetBridge")
public class WidgetBridgePlugin extends Plugin {

    @PluginMethod
    public void update(PluginCall call) {
        String data = call.getString("data");
        String config = call.getString("config");
        if (data != null) WidgetStore.saveData(getContext(), data);
        if (config != null) WidgetStore.saveConfig(getContext(), config);
        WidgetUi.refreshAll(getContext());
        call.resolve();
    }

    @PluginMethod
    public void clear(PluginCall call) {
        WidgetStore.clear(getContext());
        WidgetUi.refreshAll(getContext());
        call.resolve();
    }
}

File: ios/App/App/WidgetBridgePlugin.swift
import Capacitor
import Foundation
import WidgetKit

/// JS bridge for the home-screen widgets (see src/service/WidgetService.js).
/// Persists the task snapshot + API config in the shared App Group so the
/// widget extension can read them, then asks WidgetKit to redraw.
@objc(WidgetBridgePlugin)
public class WidgetBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WidgetBridgePlugin"
    public let jsName = "WidgetBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "update", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
    ]

    static let appGroup = "group.com.donetick.app"
    static let dataKey = "widget_tasks"
    static let configKey = "widget_config"

    @objc func update(_ call: CAPPluginCall) {
        guard let defaults = UserDefaults(suiteName: Self.appGroup) else {
            call.reject("App Group \(Self.appGroup) unavailable")
            return
        }
        if let data = call.getString("data") {
            defaults.set(data, forKey: Self.dataKey)
        }
        if let config = call.getString("config") {
            defaults.set(config, forKey: Self.configKey)
        }
        WidgetCenter.shared.reloadAllTimelines()
        call.resolve()
    }

    @objc func clear(_ call: CAPPluginCall) {
        guard let defaults = UserDefaults(suiteName: Self.appGroup) else {
            call.reject("App Group \(Self.appGroup) unavailable")
            return
        }
        defaults.removeObject(forKey: Self.dataKey)
        defaults.removeObject(forKey: Self.configKey)
        WidgetCenter.shared.reloadAllTimelines()
        call.resolve()
    }
}
